controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -123,37 +152,108 @@ func (w *Workload) ControlPlaneIsHealthy(ctx context.Context) (HealthCheckResult
 		name := node.Name
 		response[name] = nil
 
+		var owningMachine *clusterv1.Machine = nil
+		// If there is no owning Machine for the node, setting conditions are no-op.
+		for _, machine := range controlPlane.Machines {
+			if machine.Status.NodeRef == nil {
+				continue
+			}
+			if machine.Status.NodeRef.Name == node.Name {
+				owningMachine = machine
+			}
+		}
+
+		// If there is any problem with the checks below, set response to any one of the problems.
+		var anyError error = nil
 		if err := checkNodeNoExecuteCondition(node); err != nil {
-			response[name] = err
-			continue
+			anyError = err
 		}
 
-		apiServerPodKey := ctrlclient.ObjectKey{
-			Namespace: metav1.NamespaceSystem,
-			Name:      staticPodName("kube-apiserver", name),
+		// Check kube-api-server health
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeAPIServerPodNamePrefix, node, owningMachine, clusterv1.MachineAPIServerPodHealthyCondition); err != nil {
+			anyError = err
 		}
-		apiServerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, apiServerPodKey, &apiServerPod); err != nil {
-			response[name] = err
-			continue
+
+		// Check kube-controller-manager health
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeControllerManagerPodNamePrefix, node, owningMachine, clusterv1.MachineControllerManagerHealthyCondition); err != nil {
+			anyError = err
 		}
-		response[name] = checkStaticPodReadyCondition(apiServerPod)
 
-		controllerManagerPodKey := ctrlclient.ObjectKey{
-			Namespace: metav1.NamespaceSystem,
-			Name:      staticPodName("kube-controller-manager", name),
+		// Check kube-scheduler health
+		if _, err = w.checkPodStatusAndUpdateCondition(KubeSchedulerHealthyPodNamePrefix, node, owningMachine, clusterv1.MachineSchedulerPodHealthyCondition); err != nil {
+			anyError = err
 		}
-		controllerManagerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, controllerManagerPodKey, &controllerManagerPod); err != nil {
+
+		if anyError != nil {
 			response[name] = err
 			continue
 		}
-		response[name] = checkStaticPodReadyCondition(controllerManagerPod)
 	}
 
 	return response, nil
 }
 
+// checkPodStatusAndUpdateConditions returns true, only if the Pod is in Running/Ready condition.
+// There is a slight difference between returning false and error:
+// checkPodStatusAndUpdateConditions returns false, if Pod is not in Ready Condition or there is a client error.
+// checkPodStatusAndUpdateConditions returns error only when it is obvious that the Pod is not in Running/Ready state, so it does not return error on transient client errors.
+// For instance, returning (false, nil) means Pod is not ready but is not in error state as well,
+// returning (false, err) means Pod is not ready and it is in error state as well.
```

Comment:
Return a struct and an error, instead of a boolean. In most cases this will help understand the result of a function much more than having a plain boolean field